compiler/modules/front/const.bal

```diff
@@ -517,47 +520,84 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
     }
 }
 
-function foldFloatLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    // This will need to change when we support decimal
-    float|error result = floatFromDecimalLiteral(expr.untypedLiteral);
-    if result is float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+function foldFpLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError { 
+    t:Context tc = cx.typeContext();
+    var { typeSuffix, untypedLiteral, startPos } = expr;
+    float|decimal result;
+    if typeSuffix != () {
+        result = typeSuffix is s:FLOAT_TYPE_SUFFIX ? floatFromFpLiteral(untypedLiteral) : check decimalFromFpLiteral(cx, untypedLiteral, startPos);
+    }
+    else if expectedType == () || expectsFloat(tc, expectedType) {
+        result = floatFromFpLiteral(untypedLiteral);
+    }
+    else if expectsDecimal(tc, expectedType) {
+        result = check decimalFromFpLiteral(cx, untypedLiteral, startPos);
     }
     else {
-        return cx.semanticErr("invalid float literal", cause=result, pos=expr.startPos);
+        result = floatFromFpLiteral(untypedLiteral);
```

Comment:
Do we ever reach here for semantically valid input? if not let's add a comment saying semantically invalid.
(Another option is to throw an error here, but we'll have to look into if that's Ok)

Reply:
The cases like this `int i = <int>1.2;` reaches this point.
The value `1.2` is a fpLiteral and the type of this value can be either float or decimal with preference float, decimal. 
The expected type is int.
Now we need to check intesection between float and int and then decimal and int. Since both cases fail we need to consider the most prefered basic type.

Reply:
Makes sense.